test/etherscan.js

```diff
@@ -0,0 +1,369 @@
+  // Don't run if it's an external PR. Etherscan tests need API keys that can't be exposed to external PRs.
+  if (process.env.CIRCLE_PR_REPONAME !== undefined) {
+    return;
+  }
+
+  this.timeout(7000);
+  const server = new Server(CUSTOM_PORT);
+
+  before(async () => {
+    const promisified = util.promisify(server.app.listen);
+    await promisified(server.port);
+    console.log(`Server listening on port ${server.port}!`);
+  });
+
+  beforeEach(() => {
+    rimraf.sync(server.repository);
+  });
+
+  after(() => {
+    rimraf.sync(server.repository);
+  });
+
+  const assertEtherscanError = (err, res, errorMessage) => {
+    try {
+      chai.expect(res.status).to.equal(StatusCodes.BAD_REQUEST);
+      chai.expect(res.body?.error).to.equal(errorMessage);
+    } catch (e) {
+      console.log("Error: ", e);
+      console.log("Response: ", res.body);
+      throw e;
+    }
+  };
+
+  describe("Non-Session API", () => {
+    it("should fail for missing address", (done) => {
+      chai
+        .request(server.app)
+        .post("/verify/etherscan")
+        .field("chain", "1")
+        .end((err, res) => {
+          assertValidationError(err, res, "address");
+          done();
+        });
+    });
+
+    it("should fail for missing chain", (done) => {
+      chai
+        .request(server.app)
+        .post("/verify/etherscan")
+        .field("address", unusedAddress)
+        .end((err, res) => {
+          assertValidationError(err, res, "chain");
+          done();
+        });
+    });
+
+    it("should fail for invalid address", (done) => {
+      chai
+        .request(server.app)
+        .post("/verify/etherscan")
+        .field("address", invalidAddress)
+        .field("chain", "1")
+        .end((err, res) => {
+          assertValidationError(
+            err,
+            res,
+            "address",
+            `Invalid addresses: ${invalidAddress}`
+          );
+          done();
+        });
+    });
+
+    it("should fail for unsupported chain", (done) => {
+      chai
+        .request(server.app)
+        .post("/verify/etherscan")
+        .field("address", unusedAddress)
+        .field("chain", unsupportedChain)
+        .end((err, res) => {
+          assertValidationError(
+            err,
+            res,
+            "chain",
+            `Chain ${unsupportedChain} not supported!`
+          );
+          done();
+        });
+    });
+
+    it("should fail fetching a non verified contract from etherscan", (done) => {
+      chai
+        .request(server.app)
+        .post("/verify/etherscan")
+        .field("address", unusedAddress)
+        .field("chain", "1")
+        .end((err, res) => {
+          assertEtherscanError(
+            err,
+            res,
+            "This contract is not verified on Etherscan"
+          );
+          done();
+        });
+    });
+
+    it("should fail by exceeding rate limit on etherscan APIs", async () => {
```

Comment:
I would do it at the end just to be safe that then the other test succeed (I had the same problem before)